apps/server/src/modules/oauth-provider/uc/oauth-provider.client-crud.uc.spec.ts

```diff
@@ -0,0 +1,234 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { OauthProviderService } from '@shared/infra/oauth-provider/index';
+import { createMock, DeepMocked } from '@golevelup/ts-jest';
+import { OauthClient } from '@shared/infra/oauth-provider/dto/index';
+import { ICurrentUser, Permission, User } from '@shared/domain/index';
+import { AuthorizationService } from '@src/modules/index';
+import { UserRepo } from '@shared/repo/index';
+import { setupEntities, userFactory } from '@shared/testing/index';
+import { UnauthorizedException } from '@nestjs/common';
+import { MikroORM } from '@mikro-orm/core';
+import { OauthProviderClientCrudUc } from './oauth-provider.client-crud.uc';
+import resetAllMocks = jest.resetAllMocks;
+
+describe('OauthProviderUc', () => {
+	let orm: MikroORM;
+	let module: TestingModule;
+	let uc: OauthProviderClientCrudUc;
+
+	let providerService: DeepMocked<OauthProviderService>;
+	let authorizationService: DeepMocked<AuthorizationService>;
+	let userRepo: DeepMocked<UserRepo>;
+
+	let user: User;
+
+	const currentUser: ICurrentUser = { userId: 'userId' } as ICurrentUser;
+	const defaultOauthClientBody: OauthClient = {
+		scope: 'openid offline',
+		grant_types: ['authorization_code', 'refresh_token'],
+		response_types: ['code', 'token', 'id_token'],
+		redirect_uris: [],
+	};
+
+	beforeAll(async () => {
+		orm = await setupEntities();
+
+		module = await Test.createTestingModule({
+			providers: [
+				OauthProviderClientCrudUc,
+				{
+					provide: OauthProviderService,
+					useValue: createMock<OauthProviderService>(),
+				},
+				{
+					provide: AuthorizationService,
+					useValue: createMock<AuthorizationService>(),
+				},
+				{
+					provide: UserRepo,
+					useValue: createMock<UserRepo>(),
+				},
+			],
+		}).compile();
+
+		uc = module.get(OauthProviderClientCrudUc);
+		providerService = module.get(OauthProviderService);
+		authorizationService = module.get(AuthorizationService);
+		userRepo = module.get(UserRepo);
+	});
+
+	afterAll(async () => {
+		await module.close();
+		await orm.close();
+	});
+
+	beforeEach(() => {
+		user = userFactory.buildWithId();
+		userRepo.findById.mockResolvedValue(user);
+	});
+
+	afterEach(() => {
+		resetAllMocks();
+	});
+
+	describe('Client Flow', () => {
+		describe('listOAuth2Clients', () => {
```

Comment:
not a blocker for me, but again, I dislike testing multiple things in the same test. 3 expect statements per test seems almost excessive to me.

In theory, your uc does not really do anything different based on its input. I would extract one test to ensure that the permission has been checked. One test to ensure that the service gets the correct parameters, and if we want to be generous maybe another one that that also works with undefined parameters, though honestly I dont know how to break that using typescrypt. If you cant make a test fail before you implement its functionality, you dont need the test.
Finally, there would be one test remaining that the output of the service is being returned.

That way, each test is self contained, and none of them have to be changed if the usecase is extended, for example if a step  has to be added. Also, the tests would no longer suggest the usecase is responsible for listing oauth clients. Its not, its responsible for asking the service to list oauth clients, and to make sure the permission is being checked first.

Reply:
I cann see your point, we'll discuss this as a team